src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Folding/RazorCodeBlockFoldingProvider.cs

```diff
@@ -32,12 +35,71 @@ public async Task<ImmutableArray<FoldingRange>> GetFoldingRangesAsync(DocumentCo
 
                 // Directives remove the "@" but for collapsing we want to keep it for users.
                 // Shows "@code" instead of "code".
-                CollapsedText = "@" + codeBlock.DirectiveDescriptor.Directive 
+                CollapsedText = "@" + codeBlock.DirectiveDescriptor.Directive
             };
 
             builder.Add(foldingRange);
         }
 
+        var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+
+        var razorFileSyntaxWalker = new RazorFileFoldingRangeSyntaxWalker(codeDocument.Source);
+        razorFileSyntaxWalker.Visit(codeDocument.GetSyntaxTree().Root);
+        builder.AddRange(razorFileSyntaxWalker.Ranges);
+
         return builder.ToImmutableArray();
     }
+
+    private class RazorFileFoldingRangeSyntaxWalker : SyntaxWalker
+    {
+        private readonly RazorSourceDocument _source;
+        internal List<FoldingRange> Ranges { get; }
+
+        public RazorFileFoldingRangeSyntaxWalker(RazorSourceDocument source)
+        {
+            _source = source;
+            Ranges = new List<FoldingRange>();
+        }
+
+        public override void VisitRazorDirective(RazorDirectiveSyntax node)
+        {
+            // first rule: we can fold consecutive usings
+            // if we have an existing range that ends the line *before* this current using directive or begins the line *after*,
+            // we extend the range one line
+            if (SyntaxNodeUtils.IsUsingDirective(node, out _))
+            {
+                var linePosition = node.GetLinePositionSpan(_source);
+
+                var isPartOfExistingRange = false;
+                if (Ranges.LastOrDefault() is { } lastRange)
```

Comment:
I prefer avoid LINQ for these sorts of things, and would prefer an explicit count check, and index. The good news is with C# 11 list patterns avoiding linq is just as succinct as using it:

```suggestion
                if (Ranges is [.., { } lastRange])
```